### PR TITLE
Ignore ephemeral classes in test_all_present()

### LINKED ISSUE #7450: test/sql/test_compare.py::CompareAndCopyTest::test_all_present fails on armvhl, aarch64, s390x
### Describe the bug

I attempted to build SQLAlchemy 1.4.28 on Fedora 35 and Rawhide, but the builds failed because the test suite reliably failed on armvhl, aarch64, s390x in `test/sql/test_compare.py::CompareAndCopyTest::test_all_present`.

### To Reproduce

```python
# In the source tree (from tarball or the rel_1_4 branch) on a machine of an affected architecture, build sqlalchemy and run pytest twice(!) on a certain subset of tests:

[nphilipp@aarch64-test01 sqlalchemy][PROD]$ python3 setup.py build

...

[nphilipp@aarch64-test01 sqlalchemy][PROD]$ PYTHONPATH=. python3 -m pytest test/sql/test_c* -x 

... [ succeeds! ]

[nphilipp@aarch64-test01 sqlalchemy][PROD]$ PYTHONPATH=. python3 -m pytest test/sql/test_c* -x 
======================================== test session starts =========================================
platform linux -- Python 3.10.0, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/fedora/nphilipp/src/sqlalchemy, configfile: setup.cfg
plugins: xdist-2.3.0, forked-1.3.0
collected 478 items                                                                                  

test/sql/test_case_statement.py::CaseTest::test_case PASSED                                    [  0%]
test/sql/test_case_statement.py::CaseTest::test_literal_interpretation_one PASSED              [  0%]
test/sql/test_case_statement.py::CaseTest::test_literal_interpretation_three PASSED            [  0%]
test/sql/test_case_statement.py::CaseTest::test_literal_interpretation_two PASSED              [  0%]
test/sql/test_case_statement.py::CaseTest::test_literal_interpretation_two_point_five PASSED   [  1%]
test/sql/test_case_statement.py::CaseTest::test_text_doenst_explode_even_in_whenlist PASSED    [  1%]
test/sql/test_case_statement.py::CaseTest::test_text_doesnt_explode PASSED                     [  1%]
test/sql/test_case_statement.py::CaseTest::test_when_dicts[<lambda>-CASE WHEN (test.col1 = :col1_1) THEN :param_1 ELSE :param_2 END] PASSED [  1%]
test/sql/test_case_statement.py::CaseTest::test_when_dicts[<lambda>-CASE WHEN (test.col1 = :col1_1) THEN :param_1 WHEN (test.col1 = :col1_2) THEN :param_2 ELSE :param_3 END] PASSED [  1%]
test/sql/test_case_statement.py::CaseTest::test_when_dicts[<lambda>-CASE test.col1 WHEN :param_1 THEN :param_2 END] PASSED [  2%]
test/sql/test_case_statement.py::CaseTest::test_when_dicts[<lambda>-CASE test.col1 WHEN :param_1 THEN :param_2 WHEN :param_3 THEN :param_4 END] PASSED [  2%]
test/sql/test_compare.py::CacheKeyTest::test_bindparam_subclass_nocache PASSED                 [  2%]
test/sql/test_compare.py::CacheKeyTest::test_bindparam_subclass_ok_cache PASSED                [  2%]
test/sql/test_compare.py::CacheKeyTest::test_cache_key PASSED                                  [  2%]
test/sql/test_compare.py::CacheKeyTest::test_cache_key_equal PASSED                            [  3%]
test/sql/test_compare.py::CacheKeyTest::test_cache_key_no_method PASSED                        [  3%]
test/sql/test_compare.py::CacheKeyTest::test_cache_key_unknown_traverse PASSED                 [  3%]
test/sql/test_compare.py::CacheKeyTest::test_copy_internals_no_method PASSED                   [  3%]
test/sql/test_compare.py::CacheKeyTest::test_dml_not_cached_yet[dml_stmt0] PASSED              [  3%]
test/sql/test_compare.py::CacheKeyTest::test_dml_not_cached_yet[dml_stmt1] PASSED              [  4%]
test/sql/test_compare.py::CacheKeyTest::test_dml_not_cached_yet[dml_stmt2] PASSED              [  4%]
test/sql/test_compare.py::CacheKeyTest::test_generative_cache_key_regen PASSED                 [  4%]
test/sql/test_compare.py::CacheKeyTest::test_generative_cache_key_regen_w_del PASSED           [  4%]
test/sql/test_compare.py::CacheKeyTest::test_get_children_no_method PASSED                     [  5%]
test/sql/test_compare.py::CacheKeyTest::test_literal_binds PASSED                              [  5%]
test/sql/test_compare.py::CacheKeyTest::test_values_doesnt_caches_right_now PASSED             [  5%]
test/sql/test_compare.py::CompareAndCopyTest::test_all_present FAILED                          [  5%]

============================================== FAILURES ==============================================
________________________________ CompareAndCopyTest.test_all_present _________________________________
Traceback (most recent call last):
  File "/home/fedora/nphilipp/src/sqlalchemy/test/sql/test_compare.py", line 1382, in test_all_present
    is_false(bool(need), "%d Remaining classes: %r" % (len(need), need))
  File "/home/fedora/nphilipp/src/sqlalchemy/test/../lib/sqlalchemy/testing/assertions.py", line 274, in is_false
    is_(bool(a), False, msg=msg)
  File "/home/fedora/nphilipp/src/sqlalchemy/test/../lib/sqlalchemy/testing/assertions.py", line 279, in is_
    assert a is b, msg or "%r is not %r" % (a, b)
AssertionError: 1 Remaining classes: {<class 'test.sql.test_compare.CacheKeyTest.test_get_children_no_method.<locals>.Foobar2'>}
assert True is False
====================================== short test summary info =======================================
FAILED test/sql/test_compare.py::CompareAndCopyTest::test_all_present - AssertionError: 1 Remaining...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
==================================== 1 failed, 26 passed in 1.61s ====================================
[nphilipp@aarch64-test01 sqlalchemy][PROD]$
```

### Error

From the [build log](https://koji.fedoraproject.org/koji/taskinfo?taskID=79808940) for Fedora 35 on s390x:

```
...
=================================== FAILURES ===================================
_____________________ CompareAndCopyTest.test_all_present ______________________
[gw1] linux -- Python 3.10.0 /usr/bin/python3
Traceback (most recent call last):
  File "/builddir/build/BUILD/SQLAlchemy-1.4.28/test/sql/test_compare.py", line 1382, in test_all_present
    is_false(bool(need), "%d Remaining classes: %r" % (len(need), need))
  File "/builddir/build/BUILD/SQLAlchemy-1.4.28/test/../lib/sqlalchemy/testing/assertions.py", line 274, in is_false
    is_(bool(a), False, msg=msg)
  File "/builddir/build/BUILD/SQLAlchemy-1.4.28/test/../lib/sqlalchemy/testing/assertions.py", line 279, in is_
    assert a is b, msg or "%r is not %r" % (a, b)
AssertionError: 3 Remaining classes: {<class 'test.sql.test_compare.CacheKeyTest.test_get_children_no_method.<locals>.Foobar2'>, <class 'test.sql.test_compare.CacheKeyTest.test_copy_internals_no_method.<locals>.Foobar2'>, <class 'test.sql.test_compare.CacheKeyTest.test_bindparam_subclass_nocache.<locals>._literal_bindparam'>}
assert True is False
=========================== short test summary info ============================
...
``` 

### Versions

- OS: Fedora 35 and Rawhide
- Python: 3.10.0 and 3.10.1
- SQLAlchemy: 1.4.28
- Database: n/a
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): n/a

### Additional context

Failed Fedora builds:

- Rawhide [build](https://koji.fedoraproject.org/koji/buildinfo?buildID=1865779) and [corresponding task](https://koji.fedoraproject.org/koji/taskinfo?taskID=79832478)
- Fedora 35 [build](https://koji.fedoraproject.org/koji/buildinfo?buildID=1865780) and [corresponding task](https://koji.fedoraproject.org/koji/taskinfo?taskID=79808922)

I bisected the issue and this was the first commit with which I could reproduce the error:

```
commit 8d5ed4c39eed6a13397f43159278138bbe1a4a1b
Author: Mike Bayer <mike_mp@zzzcomputing.com>
Commit: Mike Bayer <mike_mp@zzzcomputing.com>

    Warn when caching is disabled / document
    ...
```

This chunk seems to be the change that triggers this behavior compared to previous versions:

```
@@ -1356,6 +1364,7 @@ class CompareAndCopyTest(CoreFixtures, fixtures.TestBase):
             and (
                 "__init__" in cls.__dict__
                 or issubclass(cls, AliasedReturnsRows)
+                or "inherit_cache" not in cls.__dict__
             )
             and not issubclass(cls, (Annotated))
             and "orm" not in cls.__module__
```

Looking at the classes reported when the test fails, they all seem to be ephemeral ones which are defined inside a test method – and this is were it gets speculative: I think these class objects still exist by the time `test_all_present()` is run, because for some reason or another they aren’t garbage-collected and therefore still show up in the subclass hierarchy of `ClauseElement` which disrupts the test. I don't know if the late garbage-collection hints at a problem in the Python interpreter on these architectures, but regardless of that, it's easy to fix that (or work around it) with a single liner fix in the test to ignore classes which are defined in a module located in `test/`, which I'll submit as a PR.

### PR DESCRIPTION
**Pull request by nphilipp**
In some circumstances, ephemeral class objects that are created within
the scope of a test method don't seem to be garbage collected directly
on exit. Filter out classes created in test modules.

Fixes: #7450

### Description

Ephemeral classes seem to leak into the class hierarchy of `ClauseElement` while `test_all_present()` is run and muck things up. This only seems to happen on armv7hl, aarch64 and 390x architectures, but there it happens reliably. The solution (or workaround, as the case may be) is to ignore classes created in the `test/` hierarchy of Python modules.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: test/sql/test_compare.py [hunk 1]
```diff
@@ -1366,6 +1366,7 @@ def test_all_present(self):
                 or "inherit_cache" not in cls.__dict__
             )
             and not issubclass(cls, (Annotated))
+            and cls.__module__.split("/", 1)[0] != "test"
```
Comment (zzzeek) [inline test/sql/test_compare.py:1369]: how about a positive test here:  `cls.__module__.startswith("sqlalchemy.") `

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 5bab1b022359387ace3f7e8f2cbd597bcc36921b of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 5bab1b022359387ace3f7e8f2cbd597bcc36921b: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

so for this one:

1. i want to see if we can do startswith("sqlalchemy.") instead
2. needs a changelog file in unreleased_14
3. then we can backport the gerrit to 1.4

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

File: test/sql/test_compare.py (hunk 1 above)
Comment (nphilipp) [reply test/sql/test_compare.py:1369]: Good call!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 135a8aaba2c6941460c7f45aa1a55c8f6b9eb43d of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 135a8aaba2c6941460c7f45aa1a55c8f6b9eb43d added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

File: test/sql/test_compare.py (hunk 1 above)
Comment (nphilipp) [reply test/sql/test_compare.py]: Works with testing for "sqlalchemy." on aarch64.

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

I've kept your change but also reverted what actually made it "break" here since I was misunderstanding what this test is trying to do.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck


that's a test artifact i've never seen before and can't reproduce locally

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

recheck


i rebuilt the PG container , not sure what's happening

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

Does not seem to have helped

Not sure what's up. I can try locally to see if I can reproduce

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

fixed the build.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3404 has been **merged**. Congratulations! :)

**Pull request closed without merging**